source/slang/slang-check-conversion.cpp

```diff
@@ -423,53 +464,211 @@ namespace Slang
+                    // 2. We have a `{}` (0 arguments), try to coerce
+                    // 3. We have a `{arg1, arg2...}`, try to coerce
 
-                    // No point in trying further if any argument fails
-                    if(!argResult)
-                        return false;
+                    auto toStructDecl = toStructDeclRef.getDecl();
+                    // 1. Ensure our `StructDecl` has declaratations created for all auto-generated constructors
+                    ensureDecl(toStructDecl, DeclCheckState::DefaultConstructorReadyForUse);
 
-                    if( coercedArg )
+                    // 2. We have a `{}` (0 arguments), try to coerce
+                    if (argCount == 0)
                     {
-                        coercedArgs.add(coercedArg);
+                        if (outToExpr)
+                        {
+                            *outToExpr = constructZeroInitListFunc(this, toStructDecl, toType, ConstructZeroInitListOptions::CheckToAvoidRecursion);
+                            (*outToExpr)->loc = fromInitializerListExpr->loc;
+                        }
+                        return true;
+                    }
+
+                    // 3. We have a `{arg1, arg2...}`, try to coerce
+                    //
+                    // For this situation we have more steps:
+                    // a. Find constructor candidate for our particular init-list
+                    // b. Collect and coerce init-list arguments into constructor parameters
+                    // c. Create InvokeExpr for our valid ConstructorDecl 
+                    List<ConstructorDecl*> ctorList = _getCtorList(this->getASTBuilder(), this, toStructDecl, nullptr);
+                    bool allowCStyleInitList = checkIfCStyleStruct(this, toStructDecl);
+
+                    List<Expr*> maybeArgList;
+                    maybeArgList.reserve(argCount);
+                    Index ioArgIndexMirror = ioArgIndex;
+                    UInt ioArgIndexCandidate = 0;
+
+                    for (auto& ctor : ctorList)
+                    {
+                        // a. Find constructor candidate for our particular init-list
+
+                        // Don't try to init default ctor with this logic
+                        auto ctorParamCount = ctor->getParameters().getCount();
+                        if (ctorParamCount == 0)
+                            continue;
+
+                        ioArgIndexCandidate = ioArgIndexMirror;
+                        ioArgIndex = ctorParamCount;
+
+                        // if allowCStyleInitList, process any ctor which comes next using the most members possible. ioArgIndex may not be 0.
+                        // if !allowCStyleInitList, process any ctor that exactly matched our argument count. ioArgIndex must start at 0.
+                        if (!allowCStyleInitList && ctorParamCount != Index(argCount))
```

Comment:
What if __init takes a variadic parameter pack?

Reply:
This requires a test.

I believe coerce logic will function correctly with type packs (meaning type packs will work as intended), but I never thought of making a test.

Reply:
It doesn't work for `variadic parameter pack`.
Test code something like this

```
struct S<T:__BuiltinArithmeticType>
{
    T v;
    __init(T x)
    {
        v = x;
    }
}
struct MyType<T:__BuiltinFloatingPointType>
{
    T val;
    T vals[10];

    __init(S<T> input)       // -> this works fine
    {
        val = input.v;
    }

    __init<each T>(expand S<each T> value)    // -> this doesn't work
    {
    }

    T getVal()
    {
        return val;
    }
}

```
diagnose:
```
error 38029: type argument 'each MyType<T>.T' does not conform to the required interface '__BuiltinArithmeticType'
    __init<each T>(expand S<each T> value)
```

Reply:
@csyonghe to take a look whether I made any syntax error for above test code.
If not, we should create a new task to accomplish this.

Reply:
This error is because `T` isn't constraint to `__BuiltinArithmeticType` as required by the interface.
You will need:
```
    __init<each T>(expand S<each T> value)
         where T : __BuiltinArithmeticType
    {
    }
```

Reply:
Also please do another merge with tot, I have recently overhauled the synthesis of generic requirements that should in theory take care of this.

Reply:
```
    __init<each T>(expand S<each T> value)
         where T : __BuiltinArithmeticType
```
this is working.

But 
```

    [mutating]
    T writeData(T value)
    {
        this.val = value;
        return value;
    }

    __init<each T>(expand S<each T> value)
         where T : __BuiltinArithmeticType
    {
        accept(expand writeData((each value).v));
    }
```
but this is wrong again:
```
error 30019: expected an expression of type 'T', got 'each T'
        accept(expand writeData((each value).v));

```